### PR TITLE
feat: add migrate command with goose

### LINKED ISSUE #44: feat: add a `migrate` target command to run database schema migrations
It would be nice if we added a `migrate` command to our CLI that runs database schema migrations so that developers can migrate up to the latest schema for a given release of OpenFGA.

```
Run database schema migrations needed for the OpenFGA server.

The migrate command is used to migrate up to a specific revision of the database schema needed for OpenFGA. If
the revision is omitted the latest or "HEAD" revision of the schema will be used.

Usage:
openfga migrate [revision]

Flags:
    --datastore-engine string - the type of database engine to run the migrations for (postgres, mysql, cockroachdb, etc..)
    --datastore-uri string        - the database connection uri (driver included) of the database to run the migrations against
```

Examples:
- `openfga migrate --datastore-engine postgres --datastore-uri postgres://localhost:5432` - migrates up to the latest "HEAD" revision of the database schema for postgres

- `openfga migrate --datastore-engine mysql --datastore-uri mysql://localhost:3306 1` - migrates up to the first revision of the database schema for mysql

I recommend we look into the https://github.com/golang-migrate/migrate package to simplify the migration execution process for a variety of database engines.

### PR DESCRIPTION
**Pull request by craigpastro**
Add a migrate command. Basically just wrapping [goose](https://github.com/pressly/goose).

```
> ./bin/openfga migrate -h
The migrate command is used to migrate the database schema needed for OpenFGA.

Usage:
  openfga migrate [flags]

Flags:
      --datastore-engine string   (required) the database engine to run the migrations for
      --datastore-uri string      (required) the connection uri of the database to run the migrations against (e.g. 'postgres://postgres:password@localhost:5432/postgres')
  -h, --help                      help for migrate
      --version int               the version to migrate to. If omitted, the latest, or "HEAD", version of the schema will be used (default -1)
```

Closes #44. 

## Review Checklist
- [x] I have added documentation for new/changed functionality in this PR or in [openfga.dev](https://github.com/openfga/openfga.dev)
- [x] The correct base branch is being used, if not `main`

### CODE REVIEW

File: pkg/cmd/root.go
Comment (craigpastro) [reply pkg/cmd/root.go:10]: When I took a look at the examples on the Cobra website I found that `Short` ones should not end with a period.

Short comes out here:
```
$ ./bin/openfga help
A high performance and flexible authorization/permission engine built for developers and inspired by Google Zanzibar.
OpenFGA is designed to make it easy for developers to model their application permissions, and to add and integrate fine-grained authorization into their applications.

Usage:
  openfga [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  help        Help about any command
  migrate     Run database schema migrations needed for the OpenFGA server
  run         Run the OpenFGA server

Flags:
  -h, --help   help for openfga

Use "openfga [command] --help" for more information about a command.
```

And Long here:
```
$ ./bin/openfga help migrate
The migrate command is used to migrate up to a specific revision of the database schema needed for OpenFGA. If the revision is omitted the latest, or "HEAD", revision of the schema will be used.

Usage:
  openfga migrate [flags] [revision]

Flags:
      --datastore-engine string   the database engine to run the migrations for: postgres
      --datastore-uri string      the connection uri of the database to run the migrations against (e.g. 'postgres://postgres:password@localhost:5432/postgres')
  -h, --help                      help for migrate
  ```

Comment (craigpastro) [reply pkg/cmd/root.go:10]: Actually, as this may take time to merge, I'll make those fixes in a different PR. Thanks!

File: go.mod
Comment (elbuo8) [inline go.mod:20]: Look at the README for the repo. It defaults to 2.7.x but we ideally want to be in 3.0 which is the actively maintained one. It will no longer have `vendored` dependencies which ultimately might make FOSSA happy :)

Comment (craigpastro) [reply go.mod]: Ah, I see. Thanks! Let me try.

Comment (craigpastro) [reply go.mod]: We are going to have to write our own migration library aren't we?

File: pkg/cmd/migrate.go
Comment (elbuo8) [inline pkg/cmd/migrate.go:44]: Inline `if`

Comment (craigpastro) [reply pkg/cmd/migrate.go:44]: This one is inlined, no?

File: Makefile
Comment (miparnisari) [inline Makefile:34]: Why did you remove the `nosemgrep` comment 😭

Comment (miparnisari) [reply Makefile:34]: You can move the URI to a variable and then put the comment above it, I suppose

File: Makefile
Comment (miparnisari) [inline Makefile:33]: shouldn't we also offer `migrate-down` in case they want to rollback?

File: Makefile
Comment (craigpastro) [reply Makefile:34]: Whoops. That must have happened when I merged main into my branch. Will fix it. Good catch. Thanks!

Comment (craigpastro) [reply Makefile:34]: And now I realized that the `run-postgres` command will no longer work since we don't have those env vars anymore.

File: pkg/cmd/migrate.go
Comment (jon-whit) [inline pkg/cmd/migrate.go:15]: Can we change this up so that there is a top-level migrations directory with subdirectories per engine? When we start adding other storage engines this will be harder to manage, because we'll have to include each directory in a COPY command to build our containers, instead of just copying the top-level migrations directory (and it's subdirectories).

File: storage/postgres/migrations/20220617_initialize_schema.sql
Comment (jon-whit) [inline storage/postgres/migrations/20220617_initialize_schema.sql:1]: Are these annotations required by goose? Can we get away without them?

File: pkg/testfixtures/storage/postgres.go
Comment (jon-whit) [inline pkg/testfixtures/storage/postgres.go:26]: Can we use `embed` here to avoid relative path issues based on where you invoke the tests from?

See 
https://pkg.go.dev/embed
and then
https://pkg.go.dev/github.com/pressly/goose/v3#SetBaseFS

File: cmd/openfga/openfga.go
Comment (craigpastro) [reply cmd/openfga/openfga.go:18]: You fixed it in your PR 🎉

File: pkg/cmd/migrate.go
Comment (craigpastro) [reply pkg/cmd/migrate.go:15]: Yep. Thanks. Done.

File: pkg/cmd/migrate.go
Comment (craigpastro) [reply pkg/cmd/migrate.go:29]: Thanks! Done.

File: storage/postgres/migrations/20220617_initialize_schema.sql
Comment (craigpastro) [reply storage/postgres/migrations/20220617_initialize_schema.sql:1]: It seems they are required. https://github.com/pressly/goose#sql-migrations I don't particularly like them either and wish they had it split into two files.

File: pkg/testfixtures/storage/postgres.go
Comment (craigpastro) [reply pkg/testfixtures/storage/postgres.go:26]: Yep! Thanks!

File: README.md
Comment (craigpastro) [reply README.md]: Updated! Thanks!

File: Makefile
Comment (craigpastro) [reply Makefile:33]: Added a `--version` flag that you can set to `0` (or any version less than the current one) that will roll you back.

File: pkg/cmd/migrate.go
Comment (craigpastro) [reply pkg/cmd/migrate.go]: Done! Thanks for the comments!

File: pkg/cmd/migrate.go
Comment (jon-whit) [inline pkg/cmd/migrate.go:33]: Can we add this panic functionality to `cmdutil` (pkg/cmd/util). For example,

```
// MustMarkFlagRequired attempts to mark the provided flag key as required and panics
// if it fails to do so.
func MustMarkFlagRequired(key string, cmd) {
    if err := cmd.MarkFlagRequired(key); err != nil {
        panic(err)
    }
}
```

or something like that..

File: go.mod
Comment (jon-whit) [inline go.mod:19]: Can we upgrade this to use the new changes that introduced the goose `NopLogger` so that we don't get log output from the goose migrations?

File: README.md
Comment (jon-whit) [inline README.md:123]: This should just be `./openfga` now.

File: go.mod
Comment (craigpastro) [reply go.mod:19]: Done. Thanks!

File: pkg/cmd/migrate.go
Comment (craigpastro) [reply pkg/cmd/migrate.go:33]: Function added. Thanks!

File: pkg/cmd/migrate.go
Comment (craigpastro) [reply pkg/cmd/migrate.go:71]: Good point!

Do you think there is any use of `version = 0` to migrate to the empty database?

File: README.md
Comment (craigpastro) [reply README.md]: Whoops. Thanks! Fixed.

File: pkg/cmd/migrate.go
Comment (craigpastro) [reply pkg/cmd/migrate.go:42]: Updated! Thank you!

**Review by elbuo8 — APPROVED** [review]

File: pkg/cmd/migrate.go
Comment (jon-whit) [inline pkg/cmd/migrate.go:55]:
```suggestion
goose.SetLogger(goose.NopLogger())
```
can we set goose to use the NopLogger so we don't get the log output from the applied migrations? If everything goes OK then we shouldn't get any log output IMO.

Comment (craigpastro) [reply pkg/cmd/migrate.go:55]: Added. I kind of like the logs to get some feedback, but I can see your point too. Thanks!

**Review by jon-whit — APPROVED** [review]
LGTM!

File: pkg/cmd/migrate.go
Comment (jon-whit) [reply pkg/cmd/migrate.go]: Probably not. Because we're always in need of revision 1 for OpenFGA to run properly 🤷 ?

**Comment by vavsab** [comment]
@craigpastro Hi there! When are you going to merge this feature? I'm going to deploy OpenFGA to my production this week. Just trying to understand whether I need to do migrations manually or better wait until this PR is merged.

**Comment by craigpastro** [comment]
Hi @vavsab! I plan to merge it today actually. Please let me know if it helps!

**Review by jon-whit — APPROVED** [review]

**Comment by vavsab** [comment]
> Hi @vavsab! I plan to merge it today actually. Please let me know if it helps!

Great! Then I will wait with deployment until merge & release. Thanks

**Pull request merged**
